File: frontend/src/types/playlist.ts
import type { Track } from './track'
import type { UserRef } from './user'
import type { PlaylistVisibility } from './friend'

export type { PlaylistVisibility }

export interface SmartPlaylistRules {
  genre?: string | null
  minimumRating?: number | null
  minimumPlayCount?: number | null
  addedWithinDays?: number | null
  limit: number
}

export interface PlaylistTrack {
  track: Track
  addedAt: string
  addedBy: UserRef
}

export interface Playlist {
  id: string
  name: string
  description: string | null
  coverUrl: string | null
  /** Legacy boolean kept for API back-compat. Prefer `visibility`. */
  isPublic: boolean
  /** Three-state visibility. Falls back to isPublic when absent. */
  visibility?: PlaylistVisibility
  /** Admin-curated featured flag for homepage / browse priority. */
  isFeatured?: boolean
  /** Manual ordering weight — lower numbers appear first when featured. */
  sortOrder?: number
  smartRules?: SmartPlaylistRules | null
  owner: UserRef
  collaborators?: UserRef[]
  /** Whether the current user is a collaborator (not the owner). */
  isCollaborator?: boolean
  tracks: PlaylistTrack[]
  followerCount: number
  totalDurationMs: number
  createdAt: string
  updatedAt: string
  isOwner?: boolean
  isSaved?: boolean
}

File: backend/src/NotSpotify.Api/Models/Playlist.cs
namespace NotSpotify.Api.Models;

public class Playlist
{
    public Guid Id { get; set; }
    public string Name { get; set; } = string.Empty;
    public string? Description { get; set; }
    public string? CoverUrl { get; set; }
    public string? CoverKey { get; set; }
    public bool IsPublic { get; set; } = true;

    /// <summary>
    /// Three-state visibility: "public" | "friends" | "private".
    /// Kept in sync with IsPublic: public=true/true, friends=false/friends, private=false/private.
    /// </summary>
    public string Visibility { get; set; } = "public";
    public bool IsFeatured { get; set; } = false;
    public int SortOrder { get; set; } = 0;
    /// <summary>
    /// JSON-serialized smart-playlist rules. Null means this is a regular,
    /// manually curated playlist.
    /// </summary>
    public string? Rules { get; set; }
    public long FollowerCount { get; set; }
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
    public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;

    public Guid OwnerId { get; set; }
    public ApplicationUser Owner { get; set; } = null!;

    public ICollection<PlaylistTrack> PlaylistTracks { get; set; } = new List<PlaylistTrack>();
    public ICollection<PlaylistMoodTag> PlaylistMoodTags { get; set; } = new List<PlaylistMoodTag>();
}
